src/types/createRouteOptions.ts

```diff
@@ -50,28 +51,40 @@ export type WithoutParent = {
 }
 
 export type WithComponent<
-  TComponent extends Component | undefined = Component | undefined
+  TComponent extends Component = Component,
+  TParams extends Record<string, unknown> = Record<string, unknown>
 > = {
   /**
    * A Vue component, which can be either synchronous or asynchronous components.
    */
-  component?: TComponent,
+  component: TComponent,
+  props?: (params: TParams) => TComponent extends Component ? MaybePromise<ComponentProps<TComponent>> : {},
 }
 
-export function isWithComponent(options: CreateRouteOptions): options is CreateRouteOptions & { component: Component } {
+export function isWithComponent(options: CreateRouteOptions): options is CreateRouteOptions & WithComponent {
   return 'component' in options && Boolean(options.component)
 }
 
 export type WithComponents<
-  TComponents extends Record<string, Component> | undefined = Record<string, Component> | undefined
+  TComponents extends Record<string, Component> = Record<string, Component>,
+  TParams extends Record<string, unknown> = Record<string, unknown>
 > = {
   /**
    * Multiple components for named views, which can be either synchronous or asynchronous components.
    */
-  components?: TComponents,
+  components: TComponents,
+  props?: {
+    [TKey in keyof TComponents]?: (params: TParams) => TComponents[TKey] extends Component ? MaybePromise<ComponentProps<TComponents[TKey]>> : {}
```

Comment:
You might think that the props type here and in `WithComponent` could be abstracted into a type that accepts the component and the params. And it totally can, but inlining these produces much better type hints. Otherwise typescript doesn't expand the type and if you get an error its more difficult to decipher. Choosing to inline this type in both so that the devx is better.